### ISSUE TITLE
jacfwd and jacrev are fundamentally broken for complex inputs

### ISSUE THREAD

**Issue by lezcano** [pytorch/pytorch#94397]
### 🐛 Describe the bug

Consider a map `f : C -> C`. `f` is called holomorphic if it's complex differentiable. By the Cauchy-Riemann equations), one shows that this is equivalent to it being real differentiable as functions `f : R^2 -> R^2` and their `2 x 2` Jacobian being representable using one complex number. Now, there are functions that are not holomorphic, the simplest of them being `x.conj()`. These are functions whose Jacobian *cannot* be represented using complex numbers. For example the Jacobian of `x.conj()` is given by the matrix `[[1, 0], [0, 1]]`. The Jacobian vector product at a vector `v \in C` is given by `v.conj()`. There is no complex number `z` such that `z*v = v.conj()`.

All this says that `jacfwd` and `jacrev` should *always* return a real Jacobian, as there is no easy test for when a function is holomorphic. More on this a possible APIs moving forward at the end though.

A few examples that are broken ATM.
```python
 x = torch.tensor(0.5+1.7j, dtype=torch.complex64)
>>> jacfwd(torch.conj)(x)  # should return tensor([[ 1,  0], [ 0, -1]])
        [ 0, -1]])
tensor(1.-0.j)
>>> jacrev(torch.conj)(x)  # should return tensor([[ 1,  0], [ 0, -1]])
tensor(1.-0.j)
>>> jacfwd(torch.abs)(x)  # should return  tensor([[0.2822, 0.9594]])
tensor(0.2822)
>>> jacrev(torch.abs)(x)   # almost! Should return tensor([[0.2822], [0.9594]])
tensor(0.2822+0.9594j)
>>> jacfwd(torch.Tensor.cfloat)(torch.ones(()))  # should return  tensor([[1, 0]])
tensor(1.+0.j)
>>> jacrev(torch.Tensor.cfloat)(torch.ones(()))   # should return  tensor([[1], [0]])
tensor(1.)
```

Note that there is no holomorphic function from `R -> C` or `C -> R` so none of the functions above are holomorphic and PyTorch should never return complex "Jacobians".

There is no easy way to test whether a function is holomorphic, but we do know that a composition of holomorphic functions is holomorphic. As such, a possible API for this would be to have a kwarg `holomorphic: bool = False`  that if set to `True` it tries to return a complex Jacobian if possible. This would be implemented by having a list of functions which are holomorphic, and making sure these are the only ones called in the computaiton of `jacfwd` / `jacrev`.

We should also add a test that tests that `jacfwd(f)(x) == jacrev(f)(x).mT`.
An issue that'd easy the implementation of this one: https://github.com/pytorch/pytorch/issues/94395

### Versions

master

**Comment by lezcano** [comment]
A fix for this issue would go as:

- Assume you have $f : C^n \to C^k$ Let's do the forward pass, the backward is analogous
- Have the function chunks blablabla return x2 the number of vectors in the complex case
- For the complex inputs, pass those vectors through view_as_complex to effectively get the 2n vectors $(1, 0, 0...), (i, 0, 0,...), (0, 1, 0...), (0, i, 0,...)$
- Run forward AD on these. You will get $2n$ vectors in $C^k$. 
- You `view_as_real` these and you end up with a $2k \times 2n$ real matrix. This is the correct result

For how to generate the basis $(1,...), (i,...)$, you can do it as above, or you can do it however you want.

**Comment by kshitij12345** [comment]
As a data-point, `jax` does forbid complex values (by default)

jacrev:
```python
import numpy as np
import jax

a = np.ones(3,)

def fn(a):
    return a * 0.5j

jax.jacrev(fn)(a)

```

Output
```
TypeError: jacrev requires real-valued outputs (output dtype that is a sub-dtype of np.floating), but got complex64. For holomorphic differentiation, pass holomorphic=True. For differentiation of non-holomorphic functions involving complex outputs, use jax.vjp directly.
```

jacfwd:
```python
import numpy as np
import jax

a = np.ones(1, dtype=np.complex64)

def fn(a):
    return jax.numpy.conjugate(a)

jax.jacfwd(fn)(a)
```

Output
```
TypeError: jacfwd requires real-valued inputs (input dtype that is a sub-dtype of np.floating), but got complex64. For holomorphic differentiation, pass holomorphic=True. For differentiation of non-holomorphic functions involving complex inputs or integer inputs, use jax.jvp directly.
```

**Comment by lezcano** [comment]
Jax's complex support is not good at all, so I don't think we should look at it too much.

About the jacfwd and jacrev check, my point is that, mathematically, those two objects are the same, so we should add a test that tests this.

Also, I think we should support jacfwd and jacrev for complex functions, for exactly the same reason we support autograd for complex functions. I think the `holomorphic` kwarg is also something worth having. If your map is holomorphic, its Jacobian has some symmetries, meaning that it has plenty of repeated information. This means two things:
- We can compute it faster, as we don't need to repeat those computations.
- The returned tensor is smaller and it already encodes all those symmetries, without incurring on numerical errors.

**Comment by albanD** [comment]
More discussion is needed to decide what needs to happen and we'll wait on @zou3519 opinion.
A small temporary fix we should do is to disable these functions on complex numbers until we resolve this discussion.

**Comment by zou3519** [comment]
Removing hi-pri since we fixed the silent correctness by raising an error (https://github.com/pytorch/pytorch/pull/94805)

**Comment by zou3519** [comment]
I caught up with @soulitzer and @albanD on the situation. I'm not an expert on complex numbers, so I'll generally defer to everyone else's thoughts on the matter.

Here is our current thinking (please let us know your thoughts, @lezcano, @kshitij12345):
- @albanD tells us that in the original complex autograd discussion, not a lot of users understood what "holomorphic" is. Also, for a `C -> C` function, returning the jacobian of a `R^2 -> R^2` function is not technically the jacobian of the `C -> C` function.
- Because of the reasons on the previous bullet, it's simpler for us to say we don't support complex functions for jacrev/jacfwd for now. We should beef up the error message to describe the various ways the user could potentially compute their jacobian (everything described in Mario's comment above) and perhaps also create a nice doc page on it and link to it.
- In the future we may see feature requests come in from users to actually handle complex jacobians a certain way, and that's when we should take a look at the use cases and consider next steps. Concretely, not a lot of users have run into our incorrect complex jacobians (e.g. torch.autograd.functional.jacobian has been wrong since the beginning of complex autograd and we have not had a bug report on it).

**Comment by lezcano** [comment]
I'm happy with just erroring out with a pretty error that points to this issue and gauge what people want to do.

Another JAX-like possibility is to implement the `holomorphic=False` parameter that JAX has. I don't know what are the exact semantics of this parameter though, as the docs are [less than good](https://jax.readthedocs.io/en/latest/_autosummary/jax.jacrev.html).

A third option would be to compute the gradient of the given function as a function from R² -> R², and then check whether the function is holomorphic (the issue here is that this is data-dependent, so perhaps we can't do it?). Then, if the function is holomorphic we return its complex Jacobian, and if it's not, we error out with a nice error message pointing to this issue.

WDYT?

**Comment by lezcano** [comment]
Also, for what is worth, I don't think that returning a real Jacobian for complex functions is an issue if this is correctly documented. We could then implement the `holomorphic=` kwarg and, in this case, if the function is complex, always return the complex Jacobian.

**Comment by lezcano** [comment]
SGTM

**Comment by zou3519** [comment]
Right so there's one thing that JAX does that we don't do right now, and that is the holomorphic flag. JAX's semantics for it are:
- if holomorphic=False, then raise an error
- if holomorphic=True, then just do what we were previously doing (create a standard basis with n (not 2n) vectors) and compute the jacobian.
There's not an easy way for a user to do the latter in PyTorch today (they'd have to manually use vjp/jvp and compose it with vmap), but hopefully we'll get some feature requests for it to motivate it.

**Comment by lezcano** [comment]
Sure, the standard basis of n vectors in `holomorphic=True` makes sense. Then, what about hiding the current behaviour behind the `holomorphic=True`, effectively implementing JAX's behaviour. Then, the ideas above could potentially be discussed and added in a `func.jac` function which could offer a better API than that of JAX as discussed in https://github.com/pytorch/pytorch/issues/94395

**Comment by albanD** [comment]
I don't think `holomorphic=` is helpful here.
If the user knows what this means, they know that asking for the jacobian for `out.real` will give them what they want (computes the same as holomorphic=True).
If the user doesn't know what it means, we might as well tell them to only give real valued function and do `out.real` to get there in most cases. This will again do the right thing.

> I don't know what are the exact semantics of this parameter though

I feel like this is a red flag! If it is not clear to you what the semantic should be. I don't expect anyone to know haha
So I don't think it is a helpful standard to follow.

**Comment by zou3519** [comment]
My opinion on holomorphic is that we should wait for a user to request the flag to do something about it, but I want to point out that there is no easy way to emulate the holomorphic flag in PyTorch today without copy/pasting the jacrev/jacfwd implementation.

That is, it's not as simple as asking for the jacobian for out.real. How would we construct a function `f` such that `torch.func.jacrev(f)(x)` gives `1.+1.6j` ?
```
In [82]: x
Out[82]: DeviceArray(0.5+0.8j, dtype=complex64, weak_type=True)

In [83]: jax.jacfwd(lambda x: x ** 2, holomorphic=True)(x)
Out[83]: DeviceArray(1.+1.6j, dtype=complex64, weak_type=True)
```

**Comment by albanD** [comment]
Getting the real part gives you what you want?
```python
>>> import torch
>>> a = torch.tensor(0.5+0.8j, requires_grad=True)
>>> torch.autograd.functional.jacobian(lambda x: (x**2).real, a)
tensor(1.-1.6000j)
```

Which is the conjugate of what you have above as we return conjugate compared to jax

**Comment by zou3519** [comment]
Yes, you're right that getting the real part gives us what we want -- sorry I should have been clearer. The point I wanted to make was that we disallow complex inputs to torch.func.{jacrev, jacfwd}, so if the user wants to use jacrev/jacfwd to actually do the computation (by modifying `f` to return the real part) they need to somehow copy-paste the code and delete that check:

```py
>>> import torch
>>> a = torch.tensor(0.5+0.8j, requires_grad=True)
>>> torch.func.jacrev(lambda x: (x**2).real)(a)
```

**Comment by albanD** [comment]
Ho yes, complex inputs should be allowed. Makes sense!

**Comment by lezcano** [comment]
I meant that it's not clear to me what it does because the JAX docs for that parameter are not great.
Now, isn't the tensor that we were returning (before deactivating these functions for complex tensor) exactly the tensors that `holomorphic=True` returns? If so, isn't it trivial to implement the `holomorphic` flag?

**Comment by lezcano** [comment]
I mean, if it were documented properly, then it should be easy to understand for anyone that knows what a holomorphic function is. Now, I would expect the intersection of people that use complex stuff and people that know what a holomorphic function is to be rather large.

We could even provide further help for the user. We could have a function `is_holomorphic(f, *args, **kwargs, atol=1e.., rtol=1e..,)` that, given a function and some valid inputs for it, it checks whether the Cauchy-Riemann equations hold at that point.

**Comment by albanD** [comment]
I agree we can do that. But my question is: why do we need this new complexity?
Asking people to get the real value gives them the same result and we don't need to introduce a brand new concept that we have to educate users about.

**Comment by lezcano** [comment]
So, there are a few things here:
- The complex Jacobian (before deactivating it throwing an error) was only correct when the function is holomorphic
- As such, if we want to provide any support to complex functions, the user has to know what a holomorphic function is. There's no other way around it.
- Now, we are currently erroring out on complex inputs so, as Richard mentioned, the `.real` trick cannot even be applied.

Which inputs do you propose we should accept / not accept?

**Comment by albanD** [comment]
> The complex Jacobian (before deactivating it throwing an error) was only correct when the function is holomorphic

Do you mean in the C -> C case or just C -> R?

> the user has to know what a holomorphic function is. There's no other way around it.

For the C -> R case, the current doc we have already covers it right?

**Comment by lezcano** [comment]
There are no non-constant holomorphic functions from C -> R, so, if a function is C -> R (or R to C for the matter), we know for sure that the Jacobian is wrong :P

**Comment by lezcano** [comment]
I mean, the Jacobian is the mathematical Jacobian, and there's no choice about what it should return. The Jacobian is the matrix representation of the differential of the function. As such, using a deliberately loose notation, we have that if the function is R^n -> R^m, then it's equal to `vmap(jvp(f), eye(n)) == vmap(vjp(f), eye(m)).mT`

In the complex case, when the function is holomorphic, the differential is a **linear** function from C^n -> C^n. As such, it can be represented by a complex matrix. This is the complex Jacobian of a holomorphic function.
 
In the rest of the cases, C^n -> R^m, R^n -> C^m or C^n -> C^m, the differential is **not** complex-linear, but only real-linear. As such, we can only have a real Jacobian though, and this is a **real** matrix of size `m x 2n`, `2m x n` or `2m x 2n` respectively.

For a slightly more formal definition using this notation, see https://github.com/pytorch/pytorch/pull/91873#pullrequestreview-1295330357 or the notes I wrote about this in section 5 of https://arxiv.org/abs/2207.06114

**Comment by lezcano** [comment]
I have not sat down and figure out the heck is going on with that Wirtinger calculus but, if I had to bet, I'd say that it's just another formalism to say exactly the same thing I said above.

My intuition behind this would be "well, there's just one kind of derivative, so the two concepts must agree" :P  I can elaborate on the reasons behind this offline if you think it'd be useful

**Comment by redwrasse** [comment]
Hello, another way to maybe move forward here. Following https://arxiv.org/pdf/1005.5170 for Wirtinger calculus.


For f: C^n -> C^m, have both W (Wirtinger derivative) and CW (conjugate Wirtinger derivative). ('treat z and z* and independent variables'). If f is holomorphic, W reduces to the standard complex derivative and CW vanishes. Similarly if f is conjugate holomorphic, CW reduces to standard conjugate derivative and W vanishes.


As I understand Pytorch uses conjugate Wirtinger derivatives (CW) which is fine. These derivative matrix entries shown above look to be CW derivatives:

```
>>> x = torch.tensor(0.5+1.7j, dtype=torch.complex64)
>>> jacfwd(torch.conj)(x)  
tensor(1.-0.j)
```

consistent with Pytorch's approach. W and CW derivatives have nice usual properties (product rule, chain rule, etc see https://arxiv.org/pdf/1005.5170). With 'treat z and z* and independent variables', chain rule looks like

 W(f comp g) = W(f) W(g) + CW(f) CW(g)

 so if we want the composition of functions to translate to matrix product of derivative matrices (D[f comp g] = Df Dg), 
 include both W and CW derivatives in the derivative matrix. ie.  for f: C^n -> C^m build a m x 2n dim matrix of complex-valued entries holding W and CW derivatives. If f is holomorphic, the CW entries will be 0 in the derivative matrix, and matrix algebra algorithms can ignore those entries. And sim. if f is conjugate holomorphic, the W entries will be zero. (so, might consider having a kwarg/flag for `homomorphic` as well as `conjugate-holomorphic` to speed things up.)

**Comment by redwrasse** [comment]
I think a derivative matrix and Hessian can be read off the Taylor expansion in terms of Wirtinger calculus (see pg. 13 https://arxiv.org/pdf/1005.5170). Taking the input as tuple (z, z*) I see first and second order terms for f: C^n -> C^m
```
derivative matrix D(f) = [W CW]

Hessian H(f) = [W^2 WCW
               CWW CW^2]

```

And these simplify for the various cases (f holomorphic -> CW=0) etc. 

I'm not so familiar with existing computational internals in Pytorch of CW and W though (https://pytorch.org/docs/stable/notes/gradcheck.html), so this may or may not be practical.
